<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Learning FourSquare API with Python</font></h1>

## Introduction

In this lab, you will learn in details how to make calls to the Foursquare API for different purposes. You will learn how to construct a URL to send a request to the API to search for a specific type of venues, to explore a particular venue, to explore a Foursquare user, to explore a geographical location, and to get trending venues around a location. Also, you will learn how to use the visualization library, Folium, to visualize the results.

## Table of Contents

1. <a href="#item1">Foursquare API Search Function</a>
2. <a href="#item2">Explore a Given Venue</a>  
3. <a href="#item3">Explore a User</a>  
4. <a href="#item4">Foursquare API Explore Function</a>  
5. <a href="#item5">Get Trending Venues</a>  

### Import necessary Libraries

In [151]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done
Solving environment: done

## Package Plan ##

  environment location: /Users/abhinavpaul/opt/anaconda3

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h0b31af3_0         1.9 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         1.9 MB

The following packages will be UPDATED:

  ca-certificates     pkgs/main::ca-certificates-2020.1.1-0 --> conda-forge::ca-certificates-2020.4.5.1-hecc5488_0
  conda                       pkgs/main::conda-4.8.3-py37_0 --> conda-forge::conda-4.8.3-py37hc8dfbb8_1

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi              pkgs/main::certifi-2020.4.5.1-py37_0 --> conda-forge::certifi-2020.4.5.1-py37hc8dfbb8_0
  openssl              pk

### Define Foursquare Credentials and Version

##### Make sure that you have created a Foursquare developer account and have your credentials handy

In [152]:
CLIENT_ID = 'TBPLKSLPDCOMMX2YIDPJ5UG2NRWHG4YOB3VXMPJI5YYP54H3' # your Foursquare ID
CLIENT_SECRET = 'OV15HD3ZH2152BWBWDWFSKBEONQJNSQCJAI0S4BO0WLFWLPE' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: TBPLKSLPDCOMMX2YIDPJ5UG2NRWHG4YOB3VXMPJI5YYP54H3
CLIENT_SECRET:OV15HD3ZH2152BWBWDWFSKBEONQJNSQCJAI0S4BO0WLFWLPE


#### Let's again assume that you are staying at the Conrad hotel. So let's start by converting the Contrad Hotel's address to its latitude and longitude coordinates.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>foursquare_agent</em>, as shown below.

In [153]:
address = '102 North End Ave, New York, NY'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

40.7149555 -74.0153365


<a id="item1"></a>

## 1. Search for a specific venue category
> `https://api.foursquare.com/v2/venues/`**search**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**`&query=`**QUERY**`&radius=`**RADIUS**`&limit=`**LIMIT**

#### Now, let's assume that it is lunch time, and you are craving Italian food. So, let's define a query to search for Italian food that is within 500 metres from the Conrad Hotel. 

In [154]:
search_query = 'Post office'
radius = 500
print(search_query + ' .... OK!')

Post office .... OK!


#### Define the corresponding URL

In [155]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=TBPLKSLPDCOMMX2YIDPJ5UG2NRWHG4YOB3VXMPJI5YYP54H3&client_secret=OV15HD3ZH2152BWBWDWFSKBEONQJNSQCJAI0S4BO0WLFWLPE&ll=40.7149555,-74.0153365&v=20180604&query=Post office&radius=500&limit=30'

#### Send the GET Request and examine the results

In [156]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ea039451a4b0a001bb64216'},
 'response': {'venues': [{'id': '4a81c3c6f964a52080f71fe3',
    'name': 'US Post Office—Church Street Station',
    'location': {'address': '90 Church St',
     'crossStreet': 'at Barclay St',
     'lat': 40.71287676753995,
     'lng': -74.01059852628421,
     'labeledLatLngs': [{'label': 'display',
       'lat': 40.71287676753995,
       'lng': -74.01059852628421}],
     'distance': 461,
     'postalCode': '10007',
     'cc': 'US',
     'city': 'New York',
     'state': 'NY',
     'country': 'United States',
     'formattedAddress': ['90 Church St (at Barclay St)',
      'New York, NY 10007',
      'United States']},
    'categories': [{'id': '4bf58dd8d48988d172941735',
      'name': 'Post Office',
      'pluralName': 'Post Offices',
      'shortName': 'Post Office',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/postoffice_',
       'suffix': '.png'},
      'primary': True}],
    'referralId':

#### Get relevant part of JSON and transform it into a *pandas* dataframe

In [157]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

id                                           name  \
0  4a81c3c6f964a52080f71fe3  US Post Office—Church Street Station            
1  4f4f9c2de4b0cad93c4b2709  US Post Office - Battery Park Retail Unit APC   
2  4cacb5232f08236aca3f8e61  USPS Post Office - Greeley Square Station       
3  4e7ce9fcd3e324c5178ca981  Scott Roen's Office                             
4  55a50090498e08199568a229  Brookfield Office Properties                    

                                                                                                                                                                                                                                               categories  \
0  [{'id': '4bf58dd8d48988d172941735', 'name': 'Post Office', 'pluralName': 'Post Offices', 'shortName': 'Post Office', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/postoffice_', 'suffix': '.png'}, 'primary': True}]              
1  [{'id': '4bf58dd8d48988d172941735', 'name': 'Post Office', 'pluralName': 'Post Offices', 'shortName': 'Post Office', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/postoffice_', 'suffix': '.png'}, 'primary': True}]              
2  [{'id': '4bf58dd8d48988d172941735', 'name': 'Post Office', 'pluralName': 'Post Offices', 'shortName': 'Post Office', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/postoffice_', 'suffix': '.png'}, 'primary': True}]              
3  []                                                                                                                                                                                                                                                       
4  [{'id': '5032885091d4c4b30a586d66', 'name': 'Real Estate Office', 'pluralName': 'Real Estate Offices', 'shortName': 'Real Estate', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/realestate_', 'suffix': '.png'}, 'primary': True}]   

     referralId  hasPerk    location.address location.crossStreet  \
0  v-1587558594  False    90 Church St        at Barclay St         
1  v-1587558594  False    15 Rector St        at Greenwich St.      
2  v-1587558594  False    31 West Broadway    NaN                   
3  v-1587558594  False    NaN                 NaN                   
4  v-1587558594  False    250 Vesey St Fl 15  NaN                   

   location.lat  location.lng  \
0  40.712877    -74.010599      
1  40.711368    -74.017019      
2  40.714349    -74.010755      
3  40.715120    -74.016020      
4  40.714704    -74.014421      

                                                       location.labeledLatLngs  \
0  [{'label': 'display', 'lat': 40.71287676753995, 'lng': -74.01059852628421}]   
1  [{'label': 'display', 'lat': 40.711368, 'lng': -74.017019}]                   
2  [{'label': 'display', 'lat': 40.71434944445903, 'lng': -74.01075486252175}]   
3  [{'label': 'display', 'lat': 40.71512, 'lng': -74.01602}]                     
4  [{'label': 'display', 'lat': 40.71470398399014, 'lng': -74.01442050933838}]   

   location.distance location.postalCode location.cc location.city  \
0  461                10007               US          New York       
1  423                10006               US          New York       
2  392                10001               US          New York       
3  60                 10282               US          New York       
4  82                 10080               US          New York       

  location.state location.country  \
0  NY             United States     
1  NY             United States     
2  NY             United States     
3  NY             United States     
4  NY             United States     

                                              location.formattedAddress  \
0  [90 Church St (at Barclay St), New York, NY 10007, United States]      
1  [15 Rector St (at Greenwich St.), New York, NY 10006, United States]   
2  [31 West Broadway, New York, NY 10001, United States]     

#### Define information of interest and filter dataframe

In [158]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name  \
0   US Post Office—Church Street Station                                      
1   US Post Office - Battery Park Retail Unit APC                             
2   USPS Post Office - Greeley Square Station                                 
3   Scott Roen's Office                                                       
4   Brookfield Office Properties                                              
5   Saks Off Fifth Corporate Office                                           
6   NYCHA (Central Office)                                                    
7   Luke's Office                                                             
8   Asphalt Green                                                             
9   Tony's Office                                                             
10  Sheindlin Law Firm                                                        
11  Duygu's Lovely Office                                                     
12  Kate's Office - MBF Asset Management                                      
13  Office of the Registrar - Borough of Manhattan Community College (BMCC)   
14  City Of New York/ Office Of Management And Budget                         
15  Robot's Post                                                              
16  HBC Saks Fifth Avenue Corporate Offices                                   
17  Dave's Office                                                             
18  GSA Federal Office Building                                               
19  st. john's university manhattan campus IT office                          
20  Mike's Office - 31st Floor                                                
21  Porter Novelli -Anthony's Office                                          
22  The People's Counsel - Law Offices of Charles L. Barberio IV              
23  E-J Remote Office Ground Zero                                             
24  Inc.                                                                      
25  NYC DoITT - 255G                                                          
26  BMCC Registrar's Office-S315                                              
27  Tribeca CERT command post                                                 
28  Ericsson WTC Office                                                       
29  Law Office of Lina M. Viviano                                             

                         categories                 address  \
0   Post Office                      90 Church St             
1   Post Office                      15 Rector St             
2   Post Office                      31 West Broadway         
3   None                             NaN                      
4   Real Estate Office               250 Vesey St Fl 15       
5   Office                           250 Vesey St             
6   Government Building              90 Church St             
7   Office                           NaN                      
8   Gym                              212 N End Ave            
9   Office                           NaN                      
10  Lawyer                           200 Vesey St             
11  Office                           NaN                      
12  Office                           1 N End Ave              
13  College Administrative Building  199 Chambers St          
14  Government Building              75 Park Pl               
15  None                             North End Ave            
16  Office                           225 Liberty St           
17  Building                         32-320                   
18  Office                           1 WTC                    
19  College Administrative Building  101 Murray St            
20  Office                           NaN                      
21  Event Space                      NaN                      
22  Miscellaneous Shop               4542 W Pine Blvd         
23  Building                         NaN                      
24  Office                           250 Green

#### Let's visualize the Italian restaurants that are nearby

In [159]:
dataframe_filtered.name

0     US Post Office—Church Street Station                                   
1     US Post Office - Battery Park Retail Unit APC                          
2     USPS Post Office - Greeley Square Station                              
3     Scott Roen's Office                                                    
4     Brookfield Office Properties                                           
5     Saks Off Fifth Corporate Office                                        
6     NYCHA (Central Office)                                                 
7     Luke's Office                                                          
8     Asphalt Green                                                          
9     Tony's Office                                                          
10    Sheindlin Law Firm                                                     
11    Duygu's Lovely Office                                                  
12    Kate's Office - MBF Asset Management                      

In [160]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the Conrad Hotel
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Conrad Hotel',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Italian restaurants as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

<a id="item2"></a>

## 2. Explore a Given Venue
> `https://api.foursquare.com/v2/venues/`**VENUE_ID**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**

### A. Let's explore the closest Italian restaurant -- _Harry's Italian Pizza Bar_

In [161]:
venue_id = '4fa862b3e4b0ebff2f749f06' # ID of Harry's Italian Pizza Bar
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/4fa862b3e4b0ebff2f749f06?client_id=TBPLKSLPDCOMMX2YIDPJ5UG2NRWHG4YOB3VXMPJI5YYP54H3&client_secret=OV15HD3ZH2152BWBWDWFSKBEONQJNSQCJAI0S4BO0WLFWLPE&v=20180604'

#### Send GET request for result

In [162]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'price', 'hasMenu', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'delivery', 'menu', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'hours', 'popular', 'seasonalHours', 'defaultHours', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto', 'colors'])


{'id': '4fa862b3e4b0ebff2f749f06',
 'name': "Harry's Italian Pizza Bar",
 'contact': {'phone': '2126081007', 'formattedPhone': '(212) 608-1007'},
 'location': {'address': '225 Murray St',
  'lat': 40.71521779064671,
  'lng': -74.01473940209351,
  'labeledLatLngs': [{'label': 'display',
    'lat': 40.71521779064671,
    'lng': -74.01473940209351},
   {'label': '?', 'lat': 40.715361, 'lng': -74.014975}],
  'postalCode': '10282',
  'cc': 'US',
  'city': 'New York',
  'state': 'NY',
  'country': 'United States',
  'formattedAddress': ['225 Murray St',
   'New York, NY 10282',
   'United States']},
 'canonicalUrl': 'https://foursquare.com/v/harrys-italian-pizza-bar/4fa862b3e4b0ebff2f749f06',
 'categories': [{'id': '4bf58dd8d48988d1ca941735',
   'name': 'Pizza Place',
   'pluralName': 'Pizza Places',
   'shortName': 'Pizza',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
    'suffix': '.png'},
   'primary': True},
  {'id': '4bf58dd8d48988d110941735',
   'name': '

### B. Get the venue's overall rating

In [163]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

7.0


That is not a very good rating. Let's check the rating of the second closest Italian restaurant.

In [164]:
venue_id = '4f3232e219836c91c7bfde94' # ID of Conca Cucina Italian Restaurant
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


Since this restaurant has no ratings, let's check the third restaurant.

In [165]:
venue_id = '3fd66200f964a520f4e41ee3' # ID of Ecco
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

7.4


Since this restaurant has a slightly better rating, let's explore it further.

### C. Get the number of tips

In [166]:
result['response']['venue']['tips']['count']

19

### D. Get the venue's tips
> `https://api.foursquare.com/v2/venues/`**VENUE_ID**`/tips?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**`&limit=`**LIMIT**

#### Create URL and send GET request. Make sure to set limit to get all tips

In [167]:
## Ecco Tips
limit = 19 # set limit to be greater than or equal to the total number of tips
url = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ea038f447b43d002388d3e2'},
 'response': {'tips': {'count': 19,
   'items': [{'id': '5ab1cb46c9a517174651d3fe',
     'createdAt': 1521601350,
     'text': 'A+ Italian food! Trust me on this: my mom’s side of the family is 100% Italian. I was born and bred to know good pasta when I see it, and Ecco is one of my all-time NYC favorites',
     'type': 'user',
     'canonicalUrl': 'https://foursquare.com/item/5ab1cb46c9a517174651d3fe',
     'lang': 'en',
     'likes': {'count': 0, 'groups': []},
     'logView': True,
     'agreeCount': 4,
     'disagreeCount': 0,
     'todo': {'count': 0},
     'user': {'id': '484542633',
      'firstName': 'Nick',
      'lastName': 'E',
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/484542633_unymNUmw_FdPs3GjXHujmHcYnN4hf8kEPADlOZuIrdcdm97VX3tFqL7fFNMNA_8Gl9NlU1GYg.jpg'}},
     'authorInteractionType': 'liked'},
    {'id': '5cb7051b180b910039f90625',
     'createdAt': 1555498267,
     

#### Get tips and list of associated features

In [168]:
tips = results['response']['tips']['items']

tip = results['response']['tips']['items'][0]
tip.keys()

dict_keys(['id', 'createdAt', 'text', 'type', 'canonicalUrl', 'lang', 'likes', 'logView', 'agreeCount', 'disagreeCount', 'todo', 'user', 'authorInteractionType'])

#### Format column width and display all tips

In [169]:
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips) # json normalize tips

# columns to keep
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display tips
tips_filtered

/Users/abhinavpaul/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  """Entry point for launching an IPython kernel.


text  \
0  A+ Italian food! Trust me on this: my mom’s side of the family is 100% Italian. I was born and bred to know good pasta when I see it, and Ecco is one of my all-time NYC favorites                      
1  Excellent food!! Osso bucco special one of the best I ever had! Lobster ravioli with porcini mushroomhomemade Italian cheesecake, tiramisu and napoleons...calamari fra diavolo was sautéed not fried   

   agreeCount  disagreeCount                        id user.firstName  \
0  4           0              5ab1cb46c9a517174651d3fe  Nick            
1  1           0              5cb7051b180b910039f90625  Lynn            

  user.lastName    user.id  
0  E             484542633  
1  B             446298346

Now remember that because we are using a personal developer account, then we can access only 2 of the restaurant's tips, instead of all 15 tips.

<a id="item3"></a>

## 3. Search a Foursquare User
> `https://api.foursquare.com/v2/users/`**USER_ID**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**

### Define URL, send GET request and display features associated with user

In [170]:
user_id = '484542633' # user ID with most agree counts and complete profile
ACCESS_TOKEN = 'WC4JYNHNPRJSZU2QOPWKCPDVOFGDIU5S2C4CPIX4GRDKZDGP'

url = 'https://api.foursquare.com/v2/users/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, ACCESS_TOKEN, VERSION) # define URL

# send GET request
results = requests.get(url).json()
user_data = results['response']['user']

# display features associated with user
user_data.keys()

dict_keys(['id', 'firstName', 'lastName', 'gender', 'canonicalUrl', 'photo', 'friends', 'tips', 'homeCity', 'bio', 'contact', 'photos', 'checkinPings', 'pings', 'type', 'mayorships', 'checkins', 'lists', 'blockedStatus', 'lenses'])

In [171]:
print('First Name: ' + user_data['firstName'])
print('Last Name: ' + user_data['lastName'])
print('Home City: ' + user_data['homeCity'])

First Name: Nick
Last Name: El-Tawil
Home City: New York, NY


#### How many tips has this user submitted?

In [172]:
user_data['tips']

{'count': 99}

Wow! So it turns out that Nick is a very active Foursquare user, with more than 250 tips.

### Get User's tips

In [173]:
# define tips URL
url = 'https://api.foursquare.com/v2/users/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

# send GET request and get user's tips
results = requests.get(url).json()
tips = results['response']['tips']['items']

# format column width
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips)

# filter columns
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display user's tips
tips_filtered

/Users/abhinavpaul/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  if __name__ == '__main__':


text  \
0   They serve coffee!!!!!!                                                                                                                                                                             
1   The linguine with clams is on point 👌                                                                                                                                                               
2   Quick, cheap lunch that tastes good! Way shorter line than Chipotle, too.                                                                                                                           
3   You’re not a real New Yorker until you’ve shame-ordered Insomnia Cookies for delivery at 3am                                                                                                        
4   Good for you yet still tasty! Clean green protein is my go-to after I hit the gym 💪                                                                                                                 
5   Burger game strong 💪                                                                                                                                                                                
6   Great burgers & fries! Also, this place is exactly what it’s like when you go to a bar in the Southwest. Source: I’m from Arizona.                                                                  
7   Açaí bowl + peanut butter + whey protein = 💪💪💪                                                                                                                                                      
8   Highly underrated and way less crowded than Central Park!                                                                                                                                           
9   Way easier to navigate than the Met proper, plus the Met Breuer focuses on modern art. If I only have a limited amount of time to spend in a museum, I would rather go here than anywhere else!     
10  Get the açaí bowl with peanut butter after your work out and thank me later 👌                                                                                                                       
11  When you want a burger, this should be the first thing that comes to mind. A+!                                                                                                                      
12  Way less crowded than Central Park! People who live in the neighborhood rave about Carl Schurz Park.                                                                                                
13  The best Mexican food in the Murray Hill / Kips Bay area!                                                                                                                                           
14  Best coffee shop in the neighborhood!                                                                                                                                                               
15  When there’s nice weather, the rooftop at Tonic East is the best place to watch the game. Perfect for March Madness & NBA finals!                                                                   
16  Rib game level 💯                                                                                                                                                                                    
17  Solid option for good Mexican food, which is hard to find in Murray Hill.                                                                                                                           
18  My parents love this place. They always go here when they visit NYC, even though I tell them there are 5,689,097,788 other pizza places. They don’t dare. That has to count for something, right?   

    agreeCount  disagreeCount                        id  
0   1           0              5accc98c0313204c9d7ec157  
1   1           0              5accbe3a911fc423730f3ed3 

#### Let's get the venue for the tip with the greatest number of agree counts

In [174]:
tip_id = '5ab5575d73fe2516ad8f363b' # tip id

# define URL
url = 'http://api.foursquare.com/v2/tips/{}?client_id={}&client_secret={}&v={}'.format(tip_id, CLIENT_ID, CLIENT_SECRET, VERSION)

# send GET Request and examine results
result = requests.get(url).json()
print(result['response']['tip']['venue']['name'])
print(result['response']['tip']['venue']['location'])

Cowgirl
{'address': '519 Hudson St', 'crossStreet': 'at W 10th St', 'lat': 40.73373338282062, 'lng': -74.0062998849649, 'labeledLatLngs': [{'label': 'display', 'lat': 40.73373338282062, 'lng': -74.0062998849649}], 'postalCode': '10014', 'cc': 'US', 'city': 'New York', 'state': 'NY', 'country': 'United States', 'formattedAddress': ['519 Hudson St (at W 10th St)', 'New York, NY 10014', 'United States']}


### Get User's friends

In [175]:
user_friends = json_normalize(user_data['friends']['groups'][0]['items'])
user_friends

Empty DataFrame
Columns: []
Index: []

Interesting. Despite being very active, it turns out that Nick does not have any friends on Foursquare. This might definitely change in the future.

### Retrieve the User's Profile Image

In [176]:
user_data

{'id': '484542633',
 'firstName': 'Nick',
 'lastName': 'El-Tawil',
 'gender': 'male',
 'canonicalUrl': 'https://foursquare.com/user/484542633',
 'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
  'suffix': '/484542633_unymNUmw_FdPs3GjXHujmHcYnN4hf8kEPADlOZuIrdcdm97VX3tFqL7fFNMNA_8Gl9NlU1GYg.jpg'},
 'friends': {'count': 0,
  'groups': [{'type': 'friends',
    'name': 'Mutual friends',
    'count': 0,
    'items': []},
   {'type': 'others', 'name': 'Other friends', 'count': 0, 'items': []}]},
 'tips': {'count': 99},
 'homeCity': 'New York, NY',
 'bio': 'https://www.tawil.team/nick-el-tawil/',
 'contact': {'twitter': 'nickeltawil'},
 'photos': {'count': 2,
  'items': [{'id': '5e41a7ce346b7c00085b0f70',
    'createdAt': 1581361102,
    'source': {'name': 'Foursquare for iOS',
     'url': 'https://foursquare.com/download/#/iphone'},
    'prefix': 'https://fastly.4sqi.net/img/general/',
    'suffix': '/484542633_cF12vc8JQJTd-hyznst-CsHHTdSoi4XwCGKhTuQXNfo.jpg',
    'width': 1440,
   

In [177]:
# 1. grab prefix of photo
# 2. grab suffix of photo
# 3. concatenate them using the image size  
Image(url='https://igx.4sqi.net/img/user/300x300/484542633_mK2Yum7T_7Tn9fWpndidJsmw2Hof_6T5vJBKCHPLMK5OL-U5ZiJGj51iwBstcpDLYa3Zvhvis.jpg')

<a id="item4"></a>

## 4. Explore a location
> `https://api.foursquare.com/v2/venues/`**explore**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**`&limit=`**LIMIT**

#### So, you just finished your gourmet dish at Ecco, and are just curious about the popular spots around the restaurant. In order to explore the area, let's start by getting the latitude and longitude values of Ecco Restaurant.

In [178]:
latitude = 40.715337
longitude = -74.008848

#### Define URL

In [179]:
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=TBPLKSLPDCOMMX2YIDPJ5UG2NRWHG4YOB3VXMPJI5YYP54H3&client_secret=OV15HD3ZH2152BWBWDWFSKBEONQJNSQCJAI0S4BO0WLFWLPE&ll=40.715337,-74.008848&v=20180604&radius=500&limit=30'

#### Send GET request and examine results

In [180]:
import requests

In [181]:
results = requests.get(url).json()
'There are {} around Ecco restaurant.'.format(len(results['response']['groups'][0]['items']))

'There are 30 around Ecco restaurant.'

#### Get relevant part of JSON

In [182]:
items = results['response']['groups'][0]['items']
items[0]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '4af5d65ff964a52091fd21e3',
  'name': 'Korin',
  'location': {'address': '57 Warren St',
   'crossStreet': 'Church St',
   'lat': 40.71482437714839,
   'lng': -74.00940425461492,
   'labeledLatLngs': [{'label': 'display',
     'lat': 40.71482437714839,
     'lng': -74.00940425461492},
    {'label': '?', 'lat': 40.714727, 'lng': -74.009399}],
   'distance': 73,
   'postalCode': '10007',
   'cc': 'US',
   'neighborhood': 'Tribeca',
   'city': 'New York',
   'state': 'NY',
   'country': 'United States',
   'formattedAddress': ['57 Warren St (Church St)',
    'New York, NY 10007',
    'United States']},
  'categories': [{'id': '4bf58dd8d48988d1f8941735',
    'name': 'Furniture / Home Store',
    'pluralName': 'Furniture / Home Stores',
    'shortName': 'Furniture / Home',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_

#### Process JSON and convert it to a clean dataframe

In [183]:
dataframe = json_normalize(items) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

name                     categories  \
0  Korin                    Furniture / Home Store          
1  Los Tacos No. 1          Taco Place                      
2  Takahachi Bakery         Bakery                          
3  Chambers Street Wines    Wine Shop                       
4  Juice Press              Vegetarian / Vegan Restaurant   
5  Heyday                   Spa                             
6  Takahachi                Sushi Restaurant                
7  Equinox Tribeca          Gym                             
8  Whole Foods Market       Grocery Store                   
9  Exceed Physical Culture  Gym / Fitness Center            

                address                     crossStreet        lat        lng  \
0  57 Warren St          Church St                       40.714824 -74.009404   
1  136 Church St         NaN                             40.714267 -74.008756   
2  25 Murray St          at Church St                    40.713653 -74.008804   
3  148 Chambers St       btwn West Broadway & Hudson St  40.715773 -74.009718   
4  83 Murray St          btwn Greenwich St & W Broadway  40.714788 -74.011132   
5  92 Reade St           NaN                             40.715726 -74.007767   
6  145 Duane St          btwn W Broadway & Church St     40.716526 -74.008101   
7  54 Murray St          at W Broadway                   40.714099 -74.009686   
8  270 Greenwich Street  at Warren St                    40.715579 -74.011368   
9  97 Reade St           bet W Broadway & Church St      40.715629 -74.007992   

                                                                                                                     labeledLatLngs  \
0  [{'label': 'display', 'lat': 40.71482437714839, 'lng': -74.00940425461492}, {'label': '?', 'lat': 40.714727, 'lng': -74.009399}]   
1  [{'label': 'display', 'lat': 40.714267, 'lng': -74.008756}]                                                                        
2  [{'label': 'display', 'lat': 40.713652845301894, 'lng': -74.0088038953017}, {'label': '?', 'lat': 40.713716, 'lng': -74.008443}]   
3  [{'label': 'display', 'lat': 40.715773063928374, 'lng': -74.00971823312332}, {'label': '?', 'lat': 40.715696, 'lng': -74.00988}]   
4  [{'label': 'display', 'lat': 40.71478769908051, 'lng': -74.0111317502157}]                                                         
5  [{'label': 'display', 'lat': 40.715726, 'lng': -74.007767}, {'label': '?', 'lat': 40.715654, 'lng': -74.00782}]                    
6  [{'label': 'display', 'lat': 40.71652647412374, 'lng': -74.00810108466207}, {'label': '?', 'lat': 40.716508, 'lng': -74.007989}]   
7  [{'label': 'display', 'lat': 40.71409860726041, 'lng': -74.0096857179283}]                                                         
8  [{'label': 'display', 'lat': 40.715579155420606, 'lng': -74.01136823958119}]                                                       
9  [{'label': 'display', 'lat': 40.7156286200256, 'lng': -74.0079922583853}, {'label': '?', 'lat': 40.715589, 'lng': -74.008105}]     

   distance postalCode  cc neighborhood      city state        country  \
0  73        10007      US  Tribeca      New York  NY    United States   
1  119       10007      US  NaN          New York  NY    United States   
2  187       10007      US  NaN          New York  NY    United States   
3  88        10007      US  NaN          New York  NY    United States   
4  202       10007      US  NaN          New York  NY    United States   
5  100       10013      US  NaN          New York  NY    United States   
6  146       10013      US  NaN          New York  NY    United States   
7  154       10007      US  NaN          New York  NY    United States   
8  214       10007      US  Tribeca      New York  NY    United States   
9  79        10013      US  Tribeca      New York  NY    United States   

                                                                        formattedAddress  \
0  [57 Warren St (Church St), New York, NY 10007, United States] 

#### Let's visualize these items on the map around our location

In [184]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


# add Ecco as a red circle mark
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='Ecco',
    fill=True,
    color='red',
    fill_color='red',
    fill_opacity=0.6
    ).add_to(venues_map)


# add popular spots to the map as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(venues_map)

# display map
venues_map

<a id="item5"></a>

## 5. Explore Trending Venues
> `https://api.foursquare.com/v2/venues/`**trending**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**

#### Now, instead of simply exploring the area around Ecco, you are interested in knowing the venues that are trending at the time you are done with your lunch, meaning the places with the highest foot traffic. So let's do that and get the trending venues around Ecco.

In [185]:
# define URL
url = 'https://api.foursquare.com/v2/venues/trending?client_id={}&client_secret={}&ll={},{}&v={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION)

# send GET request and get trending venues
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ea03890fb34b5001b3b4b77'},
 'response': {'venues': []}}

### Check if any venues are trending at this time

In [186]:
if len(results['response']['venues']) == 0:
    trending_venues_df = 'No trending venues are available at the moment!'
    
else:
    trending_venues = results['response']['venues']
    trending_venues_df = json_normalize(trending_venues)

    # filter columns
    columns_filtered = ['name', 'categories'] + ['location.distance', 'location.city', 'location.postalCode', 'location.state', 'location.country', 'location.lat', 'location.lng']
    trending_venues_df = trending_venues_df.loc[:, columns_filtered]

    # filter the category for each row
    trending_venues_df['categories'] = trending_venues_df.apply(get_category_type, axis=1)

In [187]:
# display trending venues
trending_venues_df

'No trending venues are available at the moment!'

Now, depending on when you run the above code, you might get different venues since the venues with the highest foot traffic are fetched live. 

### Visualize trending venues

In [188]:
if len(results['response']['venues']) == 0:
    venues_map = 'Cannot generate visual as no trending venues are available at the moment!'

else:
    venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


    # add Ecco as a red circle mark
    folium.features.CircleMarker(
        [latitude, longitude],
        radius=10,
        popup='Ecco',
        fill=True,
        color='red',
        fill_color='red',
        fill_opacity=0.6
    ).add_to(venues_map)


    # add the trending venues as blue circle markers
    for lat, lng, label in zip(trending_venues_df['location.lat'], trending_venues_df['location.lng'], trending_venues_df['name']):
        folium.features.CircleMarker(
            [lat, lng],
            radius=5,
            poup=label,
            fill=True,
            color='blue',
            fill_color='blue',
            fill_opacity=0.6
        ).add_to(venues_map)

In [189]:
# display map
venues_map

'Cannot generate visual as no trending venues are available at the moment!'

<a id="item6"></a>

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/). I hope you found this lab interesting and educational. Feel free to contact me if you have any questions!

This notebook is part of a course on **Coursera** called *Applied Data Science Capstone*. If you accessed this notebook outside the course, you can take this course online by clicking [here](http://cocl.us/DP0701EN_Coursera_Week2_LAB1).

<hr>
Copyright &copy; 2018 [Cognitive Class](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).